File: RootTools2/src/main/java/com/rarnu/tools/neo/data/UpdateInfo.java
package com.rarnu.tools.neo.data;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Handler;
import android.os.Message;
import com.rarnu.tools.neo.api.API;
import org.json.JSONException;
import org.json.JSONObject;

public class UpdateInfo {

    public interface UpdateInfoReadyCallback {
        void onUpdateInfoReady(UpdateInfo info);
    }

    private static UpdateInfo _instance = null;

    public String versionName = "";
    public int versionCode = 0;
    public String description = "";
    public String url = "";

    private boolean isReady = false;
    private UpdateInfoReadyCallback callback = null;

    private Handler hCallback = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if (callback != null) {
                callback.onUpdateInfoReady(_instance);
            }
            super.handleMessage(msg);
        }
    };

    private UpdateInfo(UpdateInfoReadyCallback callback) {
        _instance = this;
        this.callback = callback;
        threadInitUpdateInfo();
    }

    private void threadInitUpdateInfo() {
        isReady = false;
        new Thread(new Runnable() {
            @Override
            public void run() {
                String jsonStr = API.getUpdateInfo();
                try {
                    JSONObject json = new JSONObject(jsonStr);
                    versionCode = json.getInt("versionCode");
                    versionName = json.getString("versionName");
                    url = json.getString("url");
                    description = json.getString("description");
                } catch (JSONException e) {

                }
                isReady = true;
                hCallback.sendEmptyMessage(0);
            }
        }).start();
    }

    public boolean isNewVersion(Context ctx) {
        boolean ret = false;
        try {
            PackageManager pm = ctx.getPackageManager();
            PackageInfo pkg = pm.getPackageInfo(ctx.getPackageName(), 0);
            int verCode = pkg.versionCode;
            ret = versionCode > verCode;
        } catch (Exception e) {
        }
        return ret;
    }

    public static UpdateInfo getUpdateInfo(UpdateInfoReadyCallback callback) {
        new UpdateInfo(callback);
        return _instance;
    }

}

File: RootTools2/src/main/kotlin/com/rarnu/tools/neo/data/UpdateInfo.kt
@file:Suppress("DEPRECATION")

package com.rarnu.tools.neo.data

import android.content.Context
import com.rarnu.common.forEach
import org.json.JSONObject

class UpdateInfo {

    var versionName = ""
    var versionCode = 0
    var description = ""
    var descriptionEn = ""
    var url = ""

    fun isNewVersion(ctx: Context): Boolean {
        var ret = false
        try {
            val pm = ctx.packageManager
            val pkg = pm?.getPackageInfo(ctx.packageName, 0)
            if (pkg != null) {
                val verCode = pkg.versionCode
                ret = versionCode > verCode
            }
        } catch (e: Exception) {
        }
        return ret
    }

    companion object {

        fun fromJson(json: JSONObject): UpdateInfo? {
            var info: UpdateInfo? = null
            try {
                info = json.run { UpdateInfo().apply {
                    versionCode = getInt("versionCode")
                    versionName = getString("versionName")
                    url = getString("url")
                    description = getString("description")
                    descriptionEn = getString("desc_en")
                } }
            } catch (e: Exception) {

            }
            return info
        }

        fun listFromJson(json: JSONObject): List<UpdateInfo>? {
            var list: MutableList<UpdateInfo>? = null
            try {
                if (json.getInt("result") == 0) {
                    val arr = json.getJSONArray("data")
                    list = mutableListOf()
                    arr.forEach { _, obj ->
                        val info = fromJson(obj)
                        if (info != null) list.add(info)
                    }
                }
            } catch (e: Exception) {
            }
            return list
        }
    }

}
